### DISCUSSION TITLE
Schematics for ngMocks

### DISCUSSION THREAD

**Discussion by JsantosDK** [help-me-mom/ng-mocks#9950] [General]
### Schematics for ngMocks

**Author**: @JsantosDK 

## Goal
The primary goal is to avoid the current need to remake the initial setup that is created via the `@angular/cli`.

## Current Status:
Discussing a proposal for the schematic for Services.

## Conclusion
The MVP for this PR to be considered complete is:

 - [ ] A schematic for services;
 - [ ] A schematic for pipes;
 - [ ] A schematic for directives;
 - [ ] A schematic for components;
 - [ ] Updating the scripts to allow for the building of the schematics project;
 - [ ] Updating the build to expose the schematics;
 - [ ] Meeting quality code standards;

### Linked tickets
- PR #9945;
- Starts the development of #6293;

**Comment by JsantosDK** [comment]
## Proposal

Consider a service to be created with the name `FooBar`, then the schematics should create the following spec file:

```ts
import { MockBuilder,MockedComponentFixture, MockInstance, MockRender, } from 'ng-mocks';

import { FooBarService } from './foo-bar.service';

describe('FooBarService', () => {
  let service: FooBarService;
  let fixture: MockedComponentFixture<FooBarService>;

  MockInstance.scope();
  
  beforeEach(() => MockBuilder(FooBarService));

  beforeEach(() => {
    fixture = MockRender(FooBarService);
    service = fixture.point.componentInstance;
  })

  it('should be created', () => {
    expect(service).toBeTruthy();
  });
});
```

### Issues

1. It does not follow the current documentation on [how to write tests](https://ng-mocks.sudo.eu/extra/how-to-write-tests) or [provider of a component](https://ng-mocks.sudo.eu/guides/component-provider);
2. It assumes that `MockInstance.scope()` is always wanted;
3. It assumes `ngMocks.faster()` and `ngMocks.guts()` are never wanted;
4. It uses the `beforeEach` testing pattern, which is not the best with `ngMocks`, especially when there is a need to override the `MockInstance` of a service in a particular `it` or `describe`. The `setup` pattern or a `ngMocks` native pattern might be more appropriate;

**Comment by JsantosDK** [comment]
@satanTime any news on this?

**Reply by JsantosDK** [reply]
@satanTime any news on this?

**Reply by satanTime** [reply]
Hi, sorry. Still need to fix a19 first and other urgent issues, and then take a look at this.
The idea is awesome, and schematics should be added to the lib 100%.

Regarding how generated templates should look like, we can discuss afterwards.

**Reply by JsantosDK** [reply]
No problem, until then, I guess I can try to work around the build issues I guess.

**Comment by dmrickey** [comment]
@satanTime any news on this?
